Quantize early MIDI notes to the next row

guysv opened the pull request:
So for some time now I suspect that MIDI pattern insertions while the track is running are inaccurate. From inspecting the code I see that even with MIDI quantization being on, late notes are snapping to the current row in the pattern, even if it's a tick away from traversing to the next row.

Here's a gif demonstrating the fix, I recorded the latest schism released with my locally-compiled instance acting on the same MIDI signals:
![quantize-captions](https://user-images.githubusercontent.com/32539816/203338616-fdc5e47e-99f1-4660-ace4-0a69f479bbce.gif)

I think you'd agree that's an improvement 😄

guysv committed:
quantize early MIDI notes to the next row

guysv committed:
do not quantize midi to next row if not playing

jangler commented on schism/page_patedit.c:
```diff
@@ -2958,28 +2958,41 @@ static int patedit_record_note(song_note_t *cur_note, int channel, UNUSED int ro
+		r = (r + 1) % song_get_rows_in_pattern(current_pattern);
+		quantize_next_row = 1;
+	}
+
```
If a MIDI note is entered on the last row of a pattern and should be quantized to the next row, does it just write to the first row of the current pattern then?

guysv replied:
well.. yeah. What other behavior do you prefer?
* clamp row to max-row
* write note to next pattern

jangler replied:
Ideally it would write to the next row that would be played, which requires some nontrivial logic (which could hopefully be factored out of wherever it exists in the codebase already):

- If the current pattern is being looped, the correct behavior is to write to the first row of the current pattern
- If the song is being played normally, the correct behavior is to write to the first row of the next pattern in the order list
- Jump/break commands (Bxx, Cxx, SBx) should be correctly accounted for (in general, not just at the end of a pattern)

guysv replied:
very interesting! will work on this soon.

guysv committed:
moved row-offset shifting logic to standalone func

also pattern_editor_insert_midi now select sample/instrument to play keydown

jangler commented on schism/page_patedit.c:
```diff
@@ -2958,28 +2960,50 @@ static int patedit_record_note(song_note_t *cur_note, int channel, UNUSED int ro
 
+	/* correct late notes to the next row */
+	/* tick + 1 because processing the keydown itself takes another tick */
+	if (midi_flags & MIDI_TICK_QUANTIZE && SONG_PLAYING && !first_note
```
This the only place `first_note` is used, so making this conditional an `else if` of the previous one would be simpler and easier to understand.

guysv replied:
guess you're right. fixing.

guysv replied:
there is still a bug here. what if `midi_start_record` is zero and the song is playing? keyjazz will be quantized to next row based on tick, even if the trigger mode (ctrl-z) is none. maybe add `&& midi_start_record`

jangler commented on schism/mplink.c:
```diff
@@ -222,6 +222,29 @@ int song_find_last_channel(void)
 
 // ------------------------------------------------------------------------
 
+int song_get_pattern_offset(int * n, song_note_t ** buf, int * r, int offset)
```
I'd prefer the name `row` instead of `r` here. It's plenty short and much easier to read.

guysv replied:
fixing.

jangler commented on schism/mplink.c:
```diff
@@ -222,6 +222,29 @@ int song_find_last_channel(void)
+int song_get_pattern_offset(int * n, song_note_t ** buf, int * r, int offset)
+{
+	int len, tot;
+	if (song_get_mode() & (MODE_PATTERN_LOOP)) {
```
Not a big deal, but `MODE_PATTERN_LOOP` doesn't need to be parenthesized here.

guysv replied:
right. fixing.

jangler commented on schism/mplink.c:
```diff
@@ -222,6 +222,29 @@ int song_find_last_channel(void)
+	if (song_get_mode() & (MODE_PATTERN_LOOP)) {
+		// just wrap around current rows
+		*r = (*r + offset) % song_get_rows_in_pattern(*n);
+		return song_get_pattern(*n, buf);
```
This doesn't change the contents of `buf`, right, since we're not changing patterns? In that case you could pass `NULL` instead of `buf` here.

guysv replied:
what do you mean by
> since we're not changing patterns

this function replaces the original `song_get_pattern` call in `pattern_editor_insert_midi` to select and instantiate the pattern buffer with the right #n pattern according to the offset.

`buf` is still NULL here for that matter

jangler replied:
Never mind, I disagree with what I said originally.

jangler commented on schism/mplink.c:
```diff
@@ -222,6 +222,29 @@ int song_find_last_channel(void)
+
+	tot = song_get_rows_in_pattern(*n);
+	while (offset + *r > tot) {
+		offset - tot;
```
This statement doesn't do anything.

guysv replied:
correct. i meant `offset -= tot` fixing.

guysv committed:
jangler schismtracker/schismtracker#332 review fix

guysv committed:
jangler schismtracker/schismtracker#332 review fix

guysv committed:
Merge remote-tracking branch 'origin/midi-better-quantize' into midi-better-quantize

guysv committed:
Merge remote-tracking branch 'upstream/master' into midi-better-quantize

guysv committed:
fine-tune midi early note quantization logic

- tick 0 is not to be quantized
- `+ 1` for tick check range
  - so odd speeds will quantize more eagerly

guysv committed:
improved late note quantize check

- do not quantize if midi_start_record is false
  - to prevent quantization when not tracing song
- sample song playing start at function start
  - because of the midi playback trace assignment above
  - so starting note will not quantize
